### PR TITLE
New module = elb_application_lb

### PR DESCRIPTION
**Pull request by wimnat**
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

elb_application_lb
##### ANSIBLE VERSION

2.3
##### SUMMARY

New module for AWS application ELBs.  I have broken with naming convention compared to the classic elb module ec2_elb_lb because ELB is actually a separate part of boto3 and not part of ec2.  I suggest we rename ec2_elb_lb to elb_classic_lb.

I have also not attempted to merge classic and application load balancers in to one module because they are again separate in boto3 API and operate different enough to warrant two modules.

The module is currently unfinished.  I know there are bugs and modification of listeners doesn't work properly at the moment either but i have created the PR to get some feedback from the community as I imagine this module is probably quite sought after.

### CODE REVIEW

**Comment by wimnat** [comment]
Migrated from https://github.com/ansible/ansible-modules-extras/pull/2946

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (gundalow) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Missing full stop. `descriptions:` must be sentences.

**Review by gundalow — CHANGES_REQUESTED** [review]
needs_revision

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Done

**Comment by wimnat** [comment]
For those asking about a name change here are my reasons:

- The older "classic" elb module name of ec2_elb_lb i think is not particularly clear and I think it should be renamed to elb_classic_lb.  I have already put this forward in the proposals repo.
- Although the ELBs turn up in the EC2 section of the AWS console, they are not controlled by the ec2 API, they are controlled by a separate ELB API.  Ansible dictates that we should follow boto3 naming convention where possible so i think starting the module name with elb_ rather than ec2_ would be proper
- Someone suggested a name of ec2_alb - aside from the reason above, i think this name is ambiguous and also for the same reasons mentioned here - http://softwareengineering.stackexchange.com/questions/33349/scientific-evidence-that-supports-using-long-variable-names-instead-of-abbreviat

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Fix this name to match file name

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py:82]: It may be more obvious to name this parameter something like `public` as a boolean, but I'd be worried about AWS adding some third scheme meaning we have to obsolete the boolean and use their new value. No change required here, but would like your thoughts.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Can this default to `forward` since that's the only accepted value?

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: I think this needs to move out one indentation level to match `Field`

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Remove these single-quotes.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Can this work as an int as well, or must it be quoted?

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Please add the kwarg `exception=traceback.format_exc()` to this failure.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Can you handle NoCreds separately to raise a less generic message?

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Add `exception=traceback.format_exc()` here as well please.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Add traceback here as well (for line numbers).

**Review by ryansb — DISMISSED** [review]
We can definitely add an `elb_classic_lb` alias that will point at `ec2_elb_lb`, the renaming process takes multiple cycles because we must have one release (at least) for each phase (numbers assume starting now):

1. Allow the new name (release the alias) in 2.3 and add a warning on the old name
2. Release 2.4 and continue warning on the old name
3. Release 2.5 and rename `ec2_elb_lb` to `_ec2_elb_lb`
4. Release 2.6 without `_ec2_elb_lb`

**Comment by bgrrtt** [comment]
+1

**Comment by gundalow** [comment]
DELETED, no need to touch `CHANGELOG.md`

**Comment by ryanwalls** [comment]
Is there a related issue/PR for creating/modifying target groups?

**Comment by gundalow** [comment]
@ryanwalls https://ansible.sivel.net/pr/byfile.html is a handy way of seeing all open PRs.

**Comment by trevorr** [comment]
@wimnat Are you still working on this (and #19492)? I'd be happy to help complete the review changes if you don't have time.

I assume that requires a separate PR, so I don't want to waste anyone's time if you plan to finish this off soon. Many of us are eagerly anticipating this feature, so thanks for working on it!

**Comment by wimnat** [comment]
I'll try and look at this over the weekend

**Comment by simplesteph** [comment]
@gundalow  would it be good to remove Ansible extra and Ansible core PR from your link ? Or at least have a way to filter them out ?

**Comment by gundalow** [comment]
Cool idea, let me find out.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py:82]: I agree.  Chances of AWS creating a third schema are not low.

Also, with this scheme, at least we're following AWS API convention.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Can't default this as it's not actually an Ansilble parameter.  It's a child of `listeners`.

**Comment by simplesteph** [comment]
@wimnat @trevorr any updates? Would love to start using such a module :)

**Comment by gundalow** [comment]
FYI https://ansible.sivel.net/pr/by file . HTML has been updated to show if a PR is from the older repos

**Comment by wimnat** [comment]
It's currently failing some of my tests. I'm working on resolving this asap

On 12 Feb. 2017 8:07 pm, "John R Barker" <notifications@github.com> wrote:

FYI https://ansible.sivel.net/pr/by file . HTML has been updated to show if
a PR is from the older repos

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/ansible/ansible/pull/19491#issuecomment-279205646>, or mute
the thread

**Comment by gedl** [comment]
The documentation doesn't mention anything about specifying a region, yet the module errors if one is not specified 

example of specifying one:
```
elb_application_lb:
       name: test-alb
       region: eu-west-2
[...]
```

**Comment by wimnat** [comment]
This is standard behavior across modules.  If you set the env var
AWS_REGION then it wouldn't error but you wouldn't need to specify in the
module either.

**Comment by DanyC97** [comment]
@gundalow the url is gold mine, thanks for sharing!

**Comment by jsteinich** [comment]
The documented return values appear to be inside `load_balancer`, though the documentation makes it sound like they are top level elements.

The examples are inconsistent for casing for the listeners. It looks like they all need to be capitalized. Also, the listener protocol needs to be in `[TCP, HTTP, HTTPS]` (case sensitive).

It is unclear whether `TargetGroupArn` or `TargetGroupName` needs to be passed for the listeners.

`deletion_protection` doesn't work, though it can be passed as an attribute (`deletion_protection.enabled`)

**Comment by gedl** [comment]
It would be very useful if the module could support differential updates to the Rules, or at the very least not to touch them, only guaranteeing the actual ALB is there.

Without one of these options, it's virtually impossible to write a playbook that is safe to re-run against an existing ALB, as it will wipe all the rules, potentially containing the currently live target group.

In our particular case we aim to create a target_group per version, and the rollout process is simply to change rules orders on the ALB.  However with the current version of this file we seem to be unable to write a playbook that will 1) create a ALB if it doesn't exist, for example while provisioning a new site and 2) doesn't wipe everything from the existing ALB if it exists.

A possibility to consider ?

**Comment by wimnat** [comment]
Can you post an example of your alb task?

**Comment by gedl** [comment]
@wimnat we've worked around this by awkwardly calling the command line client and if the stdoutput does not include "LoadBalancerNotFound" then we run the elb_application_lb module, and skip otherwise. Like this:

```
- name: find relevant ALB arn
  command: "aws elbv2  describe-load-balancers --region {{region}} --names {{ env }}-{{ service }}-{{ module }}"
  register: describe_alb
  changed_when: false
  failed_when: false

- set_fact:
    alb_exists_exists: true
  when: describe_alb.stderr.find('LoadBalancerNotFound') == -1
```

The actual task executing the module looks like this:

```
- name: provisioning application load balancer
  elb_application_lb:
   name: "{{ env }}-{{ service }}-{{ module }}"
   region: "{{ region }}"
   subnets: "{{ target_subnets | default (created_subnets.results|map(attribute='subnet.id') | list) }}"
   security_groups: "{{ aws_sg.group_id }}"
   scheme: internet-facing
   tags:
      Name: "{{ env }}-{{ service }}-{{ module }}"
      Service: "{{ service }}"
      Environment: "{{ env }}"
      Module: "{{ module }}"
   listeners:
     - Protocol: HTTPS
       Port: 443
       DefaultActions:
         - Type: forward
           TargetGroupName: "{{ target_group | default (tg.target_group.target_group_name) }}"
       Certificates:
         - CertificateArn: "{{ https_certificate_arn }}"
       SslPolicy: ELBSecurityPolicy-2015-05
       Rules: "{{ alb_rules_conditions | default([]) }}"
   state: present
  register: alb
  when: (only_if_new is defined and only_if_new and alb_exists_exists is not defined) or only_if_new is not defined or not only_if_new
```

Note that the `rules` field is mandatory and the play can't know (unless awkwardly runs the aws client again) what the current rule set is, and therefore will replace the existing one with the one specified here. Same applies to the `default actions`'s target group. I'd like not to touch it if the ALB exists.

I'm open to the idea of being told I'm using the module in a completely unintended way.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py]: The handling of the exception will be identical as it just uses the message passed from AWS

**Comment by wimnat** [comment]
@gedl I guess i would say, i want Ansible to maintain state so if the Rules of the ALB do not match the task then I want Ansible to change them so that they do.  Is this not the case for you?

what if there was an `elb_application_lb_facts` module that you could use to check the existing rules?

I'm just thinking maybe you could use a facts module to get the current ALB rules and then pass these to your next task which creates ALB.  This seems counter intuitive to me though because if you're looking up the ALB for rules first then you know it exists :/

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: @naslanidis can you explain what's going on here?  I'm unsure what the convert_tg_name_arn is for - the ARN is part of listener.  When I run the module i get the following failure here:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: KeyError: 'TargetGroupName'
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_YZ6wr2/ansible_module_elb_application_lb.py\", line 713, in <module>\n    main()\n  File \"/tmp/ansible_YZ6wr2/ansible_module_elb_application_lb.py\", line 708, in main\n    create_or_update_elb(connection, connection_ec2, module)\n  File \"/tmp/ansible_YZ6wr2/ansible_module_elb_application_lb.py\", line 621, in create_or_update_elb\n    listener_changed = create_or_update_elb_listeners(connection, module, elb)\n  File \"/tmp/ansible_YZ6wr2/ansible_module_elb_application_lb.py\", line 502, in create_or_update_elb_listeners\n    default_action['TargetGroupArn'] = convert_tg_name_arn(connection, module, default_action['TargetGroupName'])  \nKeyError: 'TargetGroupName'\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 0}
```

**Comment by chrisruffalo** [comment]
@wimnat I was just messing around with this and line 496 has `module.exit_json(list=listeners)` which I'm 99% certain isn't supposed to be there unless we just want to exit early and show the listeners.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (chrisruffalo) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: I don't think this is supposed to be here.

**Comment by AndreaGiardini** [comment]
@chrisruffalo @gundalow @ryansb Sorry to bother you. Can we have more insights on what is needed for this PR and ansible/ansible#19492 to be merged? They are quite critical for whoever uses ansible with AWS

**Comment by chrisruffalo** [comment]
@AndreaGiardini I'm not seeing anything wrong with it at this point. I've been using it on something I'm doing now and it works fine but I'm not using all of the features.

**Comment by wimnat** [comment]
Still some stuff to do.

 - fixing up elb attributes to work as intended
 - fix bug in listeners code where no listeners throws exception

**Comment by wimnat** [comment]
Also, the following PR is needed for this module to function correctly. https://github.com/ansible/ansible/pull/23387

**Comment by ramireddypv** [comment]
Hi when im running this module im getting the error as 
```FAILED! => {"failed": true, "msg": "The module elb_application_lb was not found in configured module paths. Additionally, core modules are missing. If this is a checkout, run 'git pull --rebase' to correct this problem."}```

I just copied the file to the same location where other files are existing. Do I need to do anything else? Pls help with this error

**Comment by gundalow** [comment]
@ramireddypv Sounds like you've got a broken install of Ansible.

Install Ansible in the standard way then copy this file into `./library/`, 

e.g.
```
ls mystuff
inventory
roles/
library
group_vars
```

**Comment by Pallokala** [comment]
I have been now testing this for few weeks and haven't encountered any bugs.

@wimnat What is there to be fixed with the elb attributes?
I couldn't replicate the bug with the missing listeners, I tried both creating a new load balancer and configuring existing with no listeners or some listeners. There were no crashes.

**Comment by wimnat** [comment]
@Pallokala - @naslanidis added an `attributes` parameter to specify things like idle timeout however it is my intention to make these attributes first class citizens.  If you look at the module spec you can see i have specified access_logs parameters etc which are attributes.  These just need to be implemented.

This task should fail which shows you the exception i'm talking about:

```
      elb_application_lb:
        region: "{{ aws_region }}"
        name: "myelb"
        scheme: internal
        security_groups:
          - "sg-11111111"
        subnets:
         - "subnet-1111112"
         - "subnet-1111111"
        listeners:
          - Protocol: HTTP
            Port: 80
            DefaultActions:
              - Type: forward
                TargetGroupArn: "arn:aws:elasticloadbalancing:ap-southeast-2:bla"
        state: present
        tags:
          Env: "{{ env }}"
          Type: "{{ type }}"
```

I'm waiting on the PR i mention earlier to be merged and then i can complete these tasks.  The PR has two shipits so merge should happen soon.

**Comment by wimnat** [comment]
@gundalow @ryansb @willthames @s-hertel 

I think this is pretty much done now (at last!).  Final reviews welcome :)

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py:159]: you appear to have gone SnakeCase from below here.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: is it worth returning the listeners too?

**Comment by willthames** [comment]
Are there companion PRs for managing ALB target groups? Not saying it's your responsibility, just curious how you're managing them, and whether I need to worry about finding an implementation

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py:805]: Can `state` default to `present`? I usually expect this, I'm not sure if the flaw is with me or not :)

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py:159]: This is because it's passing everything directly to boto3 for listeners. I've seen similar in other modules

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Yep. Will do

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py:805]: I've been told by the core team previously that state should always be explicit.

**Comment by wimnat** [comment]
@willthames https://github.com/ansible/ansible/pull/19492 - target groups

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: This protests if I only specify TargetGroupArn and only uses TargetGroupName (to then get the ARN).

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Since these are being passed directly to boto3 we should fix the docs to reflect what is expected. Only accepts `Protocol` and `HTTP` as opposed to `protocol` and `http` right now. CamelCase all the listeners options.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: The listeners options all need to be CamelCase for the example to work.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py:744]: Not a blocker, but do we want a purge_listeners option? To do any modification to the ELB all the listeners must be specified each time.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Could add the expected keys to the error since CamelCase may be unexpected.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Is this return value unrelated to the idle_timeout option?

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Can import HAS_BOTO3 instead from ansible.module_utils.ec2

**Review by s-hertel — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (jbscare) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: The later example of the Certificates parameter is much more clear about what's expected here (a list of dicts, and not a string of the ARN, which is what the comment here made it sound like) -- it might be nice if this example matched.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (jbscare) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: The later example of the Certificates parameter is much more clear about what's expected here (a list of dicts, and not a string of the ARN, which is what the comment here made it sound like) -- it might be nice if this example matched.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (jbscare) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: The later example of the Certificates parameter is much more clear about what's expected here (a list of dicts, and not a string of the ARN, which is what the comment here made it sound like) -- it might be nice if this example matched.

Comment (jbscare) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: The later example of the Certificates parameter is much more clear about what's expected here (a list of dicts, and not a string of the ARN, which is what the comment here made it sound like) -- it might be nice if this example matched.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (jbscare) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: The later example of the Certificates parameter is much more clear about what's expected here (a list of dicts, and not a string of the ARN, which is what the comment here made it sound like) -- it might be nice if this example matched.

Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py]: done

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py]: done

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py:744]: Would prefer to do this later

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py
Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py]: done

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
~Looks great after pep8 is fixed.~ Edit: I guess there's a problem maybe with the bot allowing spaces in the docs. Hm. @gundalow Do you have thoughts on that since this is a docs issue?

shipit

**Comment by s-hertel** [comment]
@wimnat Looking forward to getting this merged. LMK if you want me to make pep8 fixes.

**Comment by wimnat** [comment]
@s-hertel i think we're there.  Once this and target_group is merged i'm going to have such a party :D

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
shipit

**Review by gundalow — APPROVED** [review]

**Review by willthames — APPROVED** [review]

**Comment by willthames** [comment]
shipit

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
In a followup PR it would be nice to wrap these calls in AWSRetry.backoff, since there are lots of calls hitting the same API in this module, so it'd be easy to hit limits at scale.

**Comment by warrd** [comment]
This PR is not python3 compatible so added a follow up #25300

**Pull request merged**
